subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java

```diff
@@ -48,35 +80,40 @@ public boolean containsValue(Object value) {
 
     @Override
     public Object get(Object key) {
+        if (key instanceof String) {
+            return attributes.get(keyFor((String) key));
+        }
         return attributes.get(key);
     }
 
     @Override
     public Object put(String key, Object value) {
         if (key == null) {
-            throw new ManifestException("The key of a manifest attribute must not be null.");
+            // NPE is the specified exception type (see Map#put javadoc)
+            throw new NullPointerException("The key of a manifest attribute must not be null.");
         }
         if (value == null) {
-            throw new ManifestException(String.format("The value of a manifest attribute must not be null (Key=%s).", key));
+            // NPE is the specified exception type (see Map#put javadoc)
+            throw new NullPointerException(String.format("The value of a manifest attribute must not be null (Key=%s).", key));
+        }
+        Name name;
+        if (key.length() > 70 || key.length() * 3 > 70 && Utf8.encodedLength(key) > 70) {
```

Comment:
❌ According to https://docs.oracle.com/javase/7/docs/api/java/util/jar/Attributes.Name.html the name is said to be non-null (which we already check for above), not longer than 70 chars (which we are now checking) and contains only ASCII ([0-9a-zA-Z_-]) chars. The check for `Utf8.encodedLength` would actually be wrong. We should make sure we respect the specs and we can look to see if `new java.util.jar.Attributes.Name(key)` below was doing the right checks so we don't need to implement anything.

Reply:
> The check for Utf8.encodedLength would actually be wrong

@lacasseio , can you please clarify?

Here's the specification:
https://docs.oracle.com/javase/7/docs/technotes/guides/jar/jar.html#Manifest_Specification

> No line may be longer than 72 bytes (not characters), in its UTF8-encoded form

> Because header names cannot be continued, the maximum length of a header name is 70 bytes (there must be a colon and a SPACE after the name).

The specification says that the maximum allowed key is 70 **bytes**
`Attributes.Name` uses `string#length`, thus it **fails** to comply with the specification.

Do you really suggest there's a bug in the specification itself?
Can you please provide the link to the **proper** specification?

Reply:
Ah, my bad here. The keys are limited to alpha+digit+_-, so string length and UTF-8 length is the same for that.

I guess I can drop Utf8.encodedLength(key) checks here.